Migrate to Spring Security 5 oauth

Brutus5000 opened the pull request:
Makes this API no longer an OAuth authorization server.
We only use the new spring-boot-starter-oauth2-resource-server.

This required a lot of changes in the whole test setup. `@WithUserDetails` & co no longer work as they require authorization required beans. Therefore everything was moved to bearer token test enhancers.

Brutus5000 committed:
Use nimbus-jose-jwt for JWT signing and verification

Prepare for removing deprecated Spring OAuth library (the new one will use nimbus too)

Brutus5000 committed:
Replace deprecated Spring MacSigner

Brutus5000 committed:
Migrate to Spring 5 security oauth

Sheikah45 commented on src/inttest/java/com/faforever/api/utils/OAuthHelper.java:
```diff
@@ -1,66 +1,67 @@
-  private OAuth2Request createOAuth2Request(Set<String> scope, Set<String> authorities) {
-    List<GrantedAuthority> grantedAuthorities = authorities == null ? null : AuthorityUtils.createAuthorityList(authorities.toArray(new String[0]));
-    return new OAuth2Request(null, "00000000-0000-0000-0000-000000000000", grantedAuthorities, true, scope, null, null, null, null);
+  /*
+
+  public RequestPostProcessor addBearerToken(
+    Integer userId,
+    String userName,
+    @NotNull Set<String> scopes,
+    @NotNull Set<String> authorities
+  ) {
+    var mappedScopes = scopes.stream().map(FafScope::new).toList();
+    var mappedRoles = scopes.stream().map(FafRole::new).toList();
+
+    return authentication(new FafAuthenticationToken(userId, mappedScopes, mappedRoles));
   }
+   */
```
Why leave commented out code?

Brutus5000 replied:
Because I played around a billion times forth and back and obviously didn't clean up afterwards :stuck_out_tongue:

bukajsytlos commented on src/inttest/java/com/faforever/api/AbstractIntegrationTest.java:
```diff
@@ -88,16 +96,17 @@ public void setUp() {
 
-  protected RequestPostProcessor getOAuthTokenWithTestUser(String scope, String authority) {
-    return getOAuthTokenWithTestUser(Set.of(scope), Set.of(authority));
+  protected RequestPostProcessor getOAuthTokenWithActiveUser(Set<String> scope, Set<String> authorities) {
```
scopes

bukajsytlos commented on src/inttest/java/com/faforever/api/AbstractIntegrationTest.java:
```diff
@@ -88,16 +96,17 @@ public void setUp() {
 
-  protected RequestPostProcessor getOAuthTokenWithTestUser(Set<String> scope, Set<String> authorities) {
-    return oAuthHelper.addBearerToken(5, "ACTIVE_USER", scope, authorities);
+  protected RequestPostProcessor getOAuthTokenForUserId(int userId, String... scope) {
```
scopes

Brutus5000 commented on src/main/java/com/faforever/api/config/security/MethodSecurityConfig.java:
```diff
@@ -0,0 +1,16 @@
+
+@Configuration
+@EnableGlobalMethodSecurity(prePostEnabled = true, securedEnabled = true)
+public class MethodSecurityConfig extends GlobalMethodSecurityConfiguration {
```
I could put this in a bean instead of a whole class :thinking:

Brutus5000 replied:
no probably not

Brutus5000 commented on src/main/java/com/faforever/api/security/AuditService.java:
```diff
@@ -16,8 +16,8 @@ public AuditService(UserSupplier userSupplier) {
-    final FafUserDetails fafUserDetails = userSupplier.get();
-    String extendedMessage = MessageFormat.format("{0} [invoked by User ''{1}'' with id ''{2}'']", message, fafUserDetails.getUsername(), fafUserDetails.getId());
+    final FafAuthenticationToken fafAuthenticationToken = userSupplier.get();
+    String extendedMessage = MessageFormat.format("{0} [invoked by User  with id ''{2}'']", message, fafAuthenticationToken.getUserId());
```
@bukajsytlos  I'm open for ideas how to keep the name but I think it would imply a db lookup fo the name everytime, which I wanted to avoid...

bukajsytlos replied:
Don't know about keep a name, but point of my comment was that since there are now only two parameters index 0 and 1, {2} would reference nonexisting value

bukajsytlos replied:
nvm, github showed me outdated code

bukajsytlos replied:
regarding name, do I understand it correctly, that name could/should be part of the user service, and it could be part of jwt token?

Brutus5000 replied:
yes we could put it into the extension

Brutus5000 replied:
Done

Brutus5000 commented on src/main/java/com/faforever/api/security/FafAuthenticationToken.java:
```diff
@@ -0,0 +1,60 @@
+      ImmutableList.<GrantedAuthority>builder()
+        .addAll(scopes)
+        .addAll(roles)
+        .add((GrantedAuthority) () -> "ROLE_USER")
```
probably needs a comment

Brutus5000 commented on src/main/java/com/faforever/api/security/FafAuthenticationToken.java:
```diff
@@ -0,0 +1,60 @@
+    this.userId = userId;
+    this.scopes = scopes;
+    this.roles = roles;
+    this.setAuthenticated(true);
```
comment here too

Brutus5000 commented on src/main/java/com/faforever/api/security/FafRole.java:
```diff
@@ -0,0 +1,23 @@
+public record FafRole(@NotNull String role) implements GrantedAuthority {
+  public static final String ROLE_PREFIX = "ROLE_";
+
+  public static boolean matches(GrantedAuthority authority, String role) {
```
can be removed

Brutus5000 commented on src/main/java/com/faforever/api/security/FafScope.java:
```diff
@@ -0,0 +1,36 @@
+public record FafScope(@NotNull String scope) implements GrantedAuthority {
+  public static final String SCOPE_PREFIX = "SCOPE_";
+
+  public static boolean matches(GrantedAuthority authority, String scope) {
```
can be removed

bukajsytlos commented on src/main/java/com/faforever/api/security/AuditService.java:
```diff
@@ -16,8 +16,8 @@ public AuditService(UserSupplier userSupplier) {
-    final FafUserDetails fafUserDetails = userSupplier.get();
-    String extendedMessage = MessageFormat.format("{0} [invoked by User ''{1}'' with id ''{2}'']", message, fafUserDetails.getUsername(), fafUserDetails.getId());
+    final FafAuthenticationToken fafAuthenticationToken = userSupplier.get();
+    String extendedMessage = MessageFormat.format("{0} [invoked by User  with id ''{2}'']", message, fafAuthenticationToken.getUserId());
```
{1}

bukajsytlos commented on src/main/java/com/faforever/api/security/elide/permission/FafUserCheck.java:
```diff
@@ -26,58 +26,54 @@ protected boolean checkOAuthScopes(OAuthScope... scope) {
 
   private boolean checkOAuthScopes(String... scope) {
```
scopes :)

bukajsytlos commented on src/main/java/com/faforever/api/security/elide/permission/FafUserCheck.java:
```diff
@@ -26,58 +26,54 @@ protected boolean checkOAuthScopes(OAuthScope... scope) {
-        log.trace("All requested permissions are granted: {}", String.join(", ", userPermissionRole));
-      } else {
-        log.debug("Permissions '{}' are not granted in requested permissions: {} , for user with id: {}", String.join(", ", missedScopes), String.join(", ", grantedUserRoles), userDetails.getId());
+  protected boolean checkUserPermission(User abstractUser, String... userPermissionRole) {
```
roles

Brutus5000 commented:
Tested on the test server now and everything seems to work (registration, vault, ...)

Brutus5000 committed:
ROLE_ prefix workaround

We want to get rid of the ROLE_ prefix in the permissions. The permissions should reflect the technical name in the database.

However, currently clients expect the ROLE_ prefix. So right now we list with and without prefix.

Brutus5000 committed:
Use the username in JWT for AuditService
